optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/drools/tri/DroolsScoringTriConstraintStream.java

```diff
@@ -0,0 +1,171 @@
+/*
+ * Copyright 2019 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.optaplanner.core.impl.score.stream.drools.tri;
+
+import java.math.BigDecimal;
+import java.util.List;
+
+import org.drools.model.Declaration;
+import org.drools.model.Global;
+import org.drools.model.PatternDSL;
+import org.drools.model.RuleItemBuilder;
+import org.drools.model.consequences.ConsequenceBuilder;
+import org.kie.api.runtime.rule.RuleContext;
+import org.optaplanner.core.api.function.ToIntTriFunction;
+import org.optaplanner.core.api.function.ToLongTriFunction;
+import org.optaplanner.core.api.function.TriFunction;
+import org.optaplanner.core.api.score.holder.AbstractScoreHolder;
+import org.optaplanner.core.impl.score.stream.drools.DroolsConstraint;
+import org.optaplanner.core.impl.score.stream.drools.DroolsConstraintFactory;
+
+public final class DroolsScoringTriConstraintStream<Solution_, A, B, C>
+        extends DroolsAbstractTriConstraintStream<Solution_, A, B, C> {
+
+    private final DroolsConstraint<Solution_> constraint;
```

Comment:
Interesting observation: none of the Drools-based CS (not uni, not bi, not tri) actually use `constraint`. Could that potentially have some implications, or can we just remove it?

Reply:
That's the magic of the kconstraint I guess :)
+0 to remove it, as you see fit